{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene Expression Studies of Response to Influenza Virus Infection Using Two Bulk RNA-Seq Datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overview"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This tutorial demonstrates how to run an RNA-Seq workflow using a *Mus musculus* dataset. Steps in the workflow include read trimming, quality control, read mapping, and counting mapped reads per gene to quantitate gene expression.\n",
    "\n",
    "All outputs used in [Tutorial 2](https://github.com/King-Laboratory/scRNASeq-miRNASeq-and-TF-Network-Analysis/blob/bda75860ace82cf180a6f9eae115ebaf2eabc5f9/Bulk_RNA-Seq_Tutorials/Bulk_RNA-Seq_Mouse/Tutorial_2_DEG_mouse.ipynb) for DEG analysis were created using this extended full dataset tutorial workflow.\n",
    "\n",
    "![Mouse Bulk RNA-seq workflow](images/Mouse_workflow.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Learning Objectives\n",
    "\n",
    "- Explore an example Bulk RNA-sequencing dataset\n",
    "- Understand the workflow of generating read counts, including:\n",
    "    - Accessing SRA metadata\n",
    "    - Quality control\n",
    "    - Adapter trimming\n",
    "    - Read mapping\n",
    "    - Counting mapped reads\n",
    "    - Quanitify gene expression levels\n",
    "- Report expression of the top 10 highly expressed genes\n",
    "- Combine read count files and store in AWS S3 bucket"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STEP 1: Getting Started"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\"> NOTE: This Jupyter Notebook was developed to run on GCP.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Without Container: Install Miniforge and Workflow Tools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Miniforge is a lightweight Conda distribution that offers a streamlined installation process and efficient package management. It provides access to a vast repository of packages.\n",
    "\n",
    "The following code performs these steps:\n",
    "- Downloads Miniforge or Mambaforge (you can use either based on preference)\n",
    "- Installs Miniforge (or Mambaforge) - no need to install conda since mamba will be available immediately\n",
    "- Using miniforge and bioconda, installs the tools that will be used in this tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">Tip: If using the Miniforge install, run the following code cells by removing the %%script false command. </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\n",
      "100 81.7M  100 81.7M    0     0   192M      0 --:--:-- --:--:-- --:--:--  192M\n",
      "14:27:47\n"
     ]
    }
   ],
   "source": [
    "# Download Miniforge or Mambaforge (you can use either based on preference)\n",
    "!curl -L -O https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-$(uname)-$(uname -m).sh\n",
    "\n",
    "# Install Miniforge (or Mambaforge) - no need to install conda since mamba will be available immediately\n",
    "!bash Miniforge3-$(uname)-$(uname -m).sh -b -u -p $HOME/miniforge > /dev/null\n",
    "!date +\"%T\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, using mambaforge and bioconda, install the tools that will be used in this tutorial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Update PATH to point to the Miniforge (or Mambaforge) bin files\n",
    "import os\n",
    "os.environ[\"PATH\"] += os.pathsep + os.environ[\"HOME\"]+\"/miniforge/bin\"\n",
    "\n",
    "#now we can easily use 'mamba' command to install software \n",
    "!mamba install -y -c conda-forge -c bioconda trimmomatic fastqc multiqc sql-magic entrez-direct gffread parallel-fastq-dump sra-tools sql-magic pyathena samtools star bowtie rsem entrez-direct subread pigz -y > /dev/null"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------------------------------------\n",
    "## If running from a container, as noted above, start with <b> STEP 2 </b> below:\n",
    "## STEP 2: Define Threads & Setup Directory Structures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify the number of available threads based on the VM. This is useful for later tools such as trimmomatic, or STAR."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads: 31\n"
     ]
    }
   ],
   "source": [
    "import multiprocessing\n",
    "import os\n",
    "\n",
    "num_cores = multiprocessing.cpu_count()\n",
    "THREADS = max(1, num_cores - 1)\n",
    "\n",
    "print(\"Number of threads:\", THREADS)\n",
    "os.environ[\"THREADS\"] = str(THREADS)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a set of directories in the sra-data-athena to store the reads, reference sequence files, and output files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jupyter/mouse-iav-infection\n"
     ]
    }
   ],
   "source": [
    "!cd $HOMEDIR\n",
    "!echo $PWD\n",
    "!mkdir -p data\n",
    "!mkdir -p data/raw_fastq\n",
    "!mkdir -p data/trimmed\n",
    "!mkdir -p data/fastqc\n",
    "!mkdir -p data/fastqc_trimmed\n",
    "!mkdir -p data/reference\n",
    "!mkdir -p data/aligned_bam\n",
    "!mkdir -p data/rsem_reference/mouse_rsem_reference\n",
    "!mkdir -p data/rsem_output\n",
    "!mkdir -p data/reference/STAR_index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STEP 3: Downloading relevant FASTQ files\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we will need to download the relevant fastq files.\n",
    "\n",
    "Because these files can be large, the process of downloading the fastq files can be quite lengthy.\n",
    "\n",
    "We will be downloading the fastq files using the SRA Toolkit from the NCBI's SRA (Sequence Read Archive). However, first we need to define the associated accession numbers in order to download."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 3.1: Finding run accession numbers.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The SRA stores sequence data in terms of runs, (SRR stands for Sequence Read Run). For the studies we are interested in, there is a single run ID for each sample. \n",
    "\n",
    "The mouse strain study has 36 samples and is represented in the Gene Expression Omnibus (GEO) and BioProject databases:\n",
    "    \n",
    "   - GEO Dataset ID: GSE66040\n",
    "    \n",
    "   - BioProject ID: PRJNA275751\n",
    "    \n",
    "The mouse lung resident cell type study has 43 samples and is represented in the Gene Expression Omnibus (GEO) and BioProject databases:\n",
    "    \n",
    "   - GEO Dataset ID: GSE165299\n",
    "    \n",
    "   - BioProject ID: PRJNA694034\n",
    "\n",
    "Since you are working in two different teams, you will analyze a subset of 8 samples each. There are files containing SRR IDs for groups of 8 samples from these two experiments. You will use one of these files for your subsequent analyses.\n",
    "\n",
    "`accs_GSE66040_set1.txt`\n",
    "                                                                                                                             \n",
    "`accs_GSE66040_set2.txt`\n",
    "                                                                                                                             \n",
    "`accs_GSE66040_set3.txt`\n",
    "                                                                                                                             \n",
    "`accs_GSE66040_set4.txt`\n",
    "                                                                                                                             \n",
    "`accs_GSE66040_set5.txt`\n",
    "\n",
    "                                                                                                                             \n",
    "`accs_GSE165299_set1.txt`\n",
    "                                                                                                                             \n",
    "`accs_GSE165299_set2.txt`\n",
    "                                                                                                                             \n",
    "`accs_GSE165299_set3.txt`\n",
    "                                                                                                                             \n",
    "`accs_GSE165299_set4.txt`\n",
    "                                                                                                                             \n",
    "`accs_GSE165299_set5.txt`\n",
    "                                                                                                                             \n",
    "`accs_GSE165299_set6.txt`\n",
    "                                                                                                                             \n",
    "You will substitute your file above (e.g., \"accs_GSE66040_set1.txt\" for \"accs.txt\") in the code blocks that follow."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 3.2 Downloading multiple files using the SRA-toolkit."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The code uses prefetch to download multiple SRA files in parallel. It reads the list of SRR IDs from accs.txt, uses xargs to execute prefetch for each ID, and specifies the output directory and the -f option to create FASTQ files in the same directory as the SRA files. To speed up the download the code uses -P $THREADS option allowing parallel execution using the specified number of threads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Resolving 'SRR1810156'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Resolving 'SRR1810158'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Resolving 'SRR1810163'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Resolving 'SRR1810167'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Resolving 'SRR1810188'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Resolving 'SRR1810171'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Resolving 'SRR1810162'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Resolving 'SRR1810176'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Downloading 'SRR1810163'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  SRA Normalized Format file is being retrieved\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  Downloading via HTTPS...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Downloading 'SRR1810162'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  SRA Normalized Format file is being retrieved\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  Downloading via HTTPS...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Downloading 'SRR1810156'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  SRA Normalized Format file is being retrieved\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  Downloading via HTTPS...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Downloading 'SRR1810158'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  SRA Normalized Format file is being retrieved\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  Downloading via HTTPS...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Downloading 'SRR1810171'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  SRA Normalized Format file is being retrieved\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  Downloading via HTTPS...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Downloading 'SRR1810188'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  SRA Normalized Format file is being retrieved\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  Downloading via HTTPS...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Downloading 'SRR1810176'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  SRA Normalized Format file is being retrieved\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  Downloading via HTTPS...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1: 1) Downloading 'SRR1810167'...\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  SRA Normalized Format file is being retrieved\n",
      "2026-01-08T19:46:04 prefetch.3.2.1:  Downloading via HTTPS...\n",
      "2026-01-08T19:46:17 prefetch.3.2.1:  HTTPS download succeed\n",
      "2026-01-08T19:46:19 prefetch.3.2.1:  'SRR1810156' is valid: 689203797 bytes were streamed from 689192467\n",
      "2026-01-08T19:46:19 prefetch.3.2.1: 1) 'SRR1810156' was downloaded successfully\n",
      "2026-01-08T19:46:19 prefetch.3.2.1: 1) Resolving 'SRR1810156's dependencies...\n",
      "2026-01-08T19:46:19 prefetch.3.2.1: 'SRR1810156' has 0 dependencies\n",
      "2026-01-08T19:46:22 prefetch.3.2.1:  HTTPS download succeed\n",
      "2026-01-08T19:46:24 prefetch.3.2.1:  'SRR1810171' is valid: 858051887 bytes were streamed from 858041466\n",
      "2026-01-08T19:46:24 prefetch.3.2.1: 1) 'SRR1810171' was downloaded successfully\n",
      "2026-01-08T19:46:24 prefetch.3.2.1: 1) Resolving 'SRR1810171's dependencies...\n",
      "2026-01-08T19:46:24 prefetch.3.2.1:  HTTPS download succeed\n",
      "2026-01-08T19:46:24 prefetch.3.2.1: 'SRR1810171' has 0 dependencies\n",
      "2026-01-08T19:46:25 prefetch.3.2.1:  HTTPS download succeed\n",
      "2026-01-08T19:46:26 prefetch.3.2.1:  HTTPS download succeed\n",
      "2026-01-08T19:46:26 prefetch.3.2.1:  'SRR1810162' is valid: 937641943 bytes were streamed from 937631402\n",
      "2026-01-08T19:46:26 prefetch.3.2.1: 1) 'SRR1810162' was downloaded successfully\n",
      "2026-01-08T19:46:26 prefetch.3.2.1: 1) Resolving 'SRR1810162's dependencies...\n",
      "2026-01-08T19:46:26 prefetch.3.2.1: 'SRR1810162' has 0 dependencies\n",
      "2026-01-08T19:46:27 prefetch.3.2.1:  HTTPS download succeed\n",
      "2026-01-08T19:46:27 prefetch.3.2.1:  'SRR1810188' is valid: 927545665 bytes were streamed from 927539466\n",
      "2026-01-08T19:46:27 prefetch.3.2.1: 1) 'SRR1810188' was downloaded successfully\n",
      "2026-01-08T19:46:27 prefetch.3.2.1: 1) Resolving 'SRR1810188's dependencies...\n",
      "2026-01-08T19:46:28 prefetch.3.2.1: 'SRR1810188' has 0 dependencies\n",
      "2026-01-08T19:46:28 prefetch.3.2.1:  HTTPS download succeed\n",
      "2026-01-08T19:46:28 prefetch.3.2.1:  'SRR1810158' is valid: 963995190 bytes were streamed from 963989466\n",
      "2026-01-08T19:46:28 prefetch.3.2.1: 1) 'SRR1810158' was downloaded successfully\n",
      "2026-01-08T19:46:28 prefetch.3.2.1: 1) Resolving 'SRR1810158's dependencies...\n",
      "2026-01-08T19:46:28 prefetch.3.2.1: 'SRR1810158' has 0 dependencies\n",
      "2026-01-08T19:46:29 prefetch.3.2.1:  HTTPS download succeed\n",
      "2026-01-08T19:46:29 prefetch.3.2.1:  'SRR1810167' is valid: 1031535353 bytes were streamed from 1031525465\n",
      "2026-01-08T19:46:29 prefetch.3.2.1: 1) 'SRR1810167' was downloaded successfully\n",
      "2026-01-08T19:46:29 prefetch.3.2.1: 1) Resolving 'SRR1810167's dependencies...\n",
      "2026-01-08T19:46:29 prefetch.3.2.1: 'SRR1810167' has 0 dependencies\n",
      "2026-01-08T19:46:30 prefetch.3.2.1:  'SRR1810163' is valid: 980143022 bytes were streamed from 980135470\n",
      "2026-01-08T19:46:30 prefetch.3.2.1: 1) 'SRR1810163' was downloaded successfully\n",
      "2026-01-08T19:46:30 prefetch.3.2.1: 1) Resolving 'SRR1810163's dependencies...\n",
      "2026-01-08T19:46:30 prefetch.3.2.1: 'SRR1810163' has 0 dependencies\n",
      "2026-01-08T19:46:31 prefetch.3.2.1:  'SRR1810176' is valid: 1066064511 bytes were streamed from 1066049465\n",
      "2026-01-08T19:46:31 prefetch.3.2.1: 1) 'SRR1810176' was downloaded successfully\n",
      "2026-01-08T19:46:31 prefetch.3.2.1: 1) Resolving 'SRR1810176's dependencies...\n",
      "2026-01-08T19:46:31 prefetch.3.2.1: 'SRR1810176' has 0 dependencies\n"
     ]
    }
   ],
   "source": [
    "!cat accs_GSE66040_set1.txt | xargs -P $THREADS -I {} prefetch {} -O data/raw_fastq -f yes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 3.3 Converting Multiple SRA files to Fastq\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this step, the SRA files will be processed in parallel using parallel-fastq-dump. Each SRR ID from accs.txt will be read, and xargs will be used to execute parallel-fastq-dump for each SRA ID. This will result in the creation of two paired-end FASTQ files for each SRR ID, which will be compressed into a .gz file to save space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spots read      : 17,088,302\n",
      "reads read      : 17,088,302\n",
      "reads written   : 17,088,302\n",
      "spots read      : 23,414,444\n",
      "reads read      : 23,414,444\n",
      "reads written   : 23,414,444\n",
      "spots read      : 22,942,035\n",
      "reads read      : 22,942,035\n",
      "reads written   : 22,942,035\n",
      "spots read      : 23,907,580\n",
      "reads read      : 23,907,580\n",
      "reads written   : 23,907,580\n",
      "spots read      : 25,525,585\n",
      "reads read      : 25,525,585\n",
      "reads written   : 25,525,585\n",
      "spots read      : 20,661,559\n",
      "reads read      : 20,661,559\n",
      "reads written   : 20,661,559\n",
      "spots read      : 26,138,852\n",
      "reads read      : 26,138,852\n",
      "reads written   : 26,138,852\n",
      "spots read      : 24,241,421\n",
      "reads read      : 24,241,421\n",
      "reads written   : 24,241,421\n"
     ]
    }
   ],
   "source": [
    "#process with parallel-fastq-dump using piping\n",
    "#!cat accs_GSE66040_set1.txt | xargs -I {} parallel-fastq-dump -O data/raw_fastq/ --tmpdir . --threads $THREADS --gzip --split-files --sra-id {}\n",
    "!cat accs_GSE66040_set1.txt | xargs -I {} fasterq-dump -O data/raw_fastq/ -t . {}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As before, it is good practice to turn .fastq files into .fastq.gz files to save space.\n",
    "\n",
    "In our case, we will actually need to concatenate the fastq files later on, and so will zip after this.\n",
    "\n",
    "The non-redundant SRA files can also be deleted to save more space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "find: ‚Äòdata/raw_fastq/SRR1810171‚Äô: No such file or directory\n",
      "find: ‚Äòdata/raw_fastq/SRR1810162‚Äô: No such file or directory\n",
      "find: ‚Äòdata/raw_fastq/SRR1810176‚Äô: No such file or directory\n",
      "find: ‚Äòdata/raw_fastq/SRR1810188‚Äô: No such file or directory\n",
      "find: ‚Äòdata/raw_fastq/SRR1810158‚Äô: No such file or directory\n",
      "find: ‚Äòdata/raw_fastq/SRR1810167‚Äô: No such file or directory\n",
      "find: ‚Äòdata/raw_fastq/SRR1810163‚Äô: No such file or directory\n",
      "find: ‚Äòdata/raw_fastq/SRR1810156‚Äô: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "#find and delete all SRR subfolders in the raw_fastq directory\n",
    "!find data/raw_fastq -type d -name 'SRR*' -exec rm -rf {} \\;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 3.6 Download reference transcriptome files that will be used by STAR\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This step downloads and prepares the reference data needed for your RNA-Seq analysis. It retrieves three essential files:\n",
    "\n",
    "- **Mouse genome (Mus_musculus.GRCm39.dna.primary_assembly.fa.gz)**: This compressed FASTA file contains the complete mouse genome sequence, that will be used as the reference for aligning your RNA-seq reads.\n",
    "- **Mouse gene annotations (Mus_musculus.GRCm39.115.gtf.gz)**: This compressed GTF file provides information about the genes and transcripts in the mouse genome, including their locations and structures. This data will crucial for interpreting the aligned RNA-Seq reads and understanding what genes are expressed in each.\n",
    "- **Mouse feature table (GCF_000001635.27_GRCm39_feature_table.txt.gz)**: This compressed table provides additional annotations for the mouse genome features, potentially including information about gene functions and pathways. This step will further used to analyze the differential gene expression (DEG) analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2026-01-08 16:32:30--  ftp://ftp.ensembl.org/pub/release-115/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.primary_assembly.fa.gz\n",
      "           => ‚Äòdata/reference/mouse_genome.fa.gz‚Äô\n",
      "Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.193.169\n",
      "Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.193.169|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /pub/release-115/fasta/mus_musculus/dna ... done.\n",
      "==> SIZE Mus_musculus.GRCm39.dna.primary_assembly.fa.gz ... 806418890\n",
      "==> PASV ... done.    ==> RETR Mus_musculus.GRCm39.dna.primary_assembly.fa.gz ... done.\n",
      "Length: 806418890 (769M) (unauthoritative)\n",
      "\n",
      "Mus_musculus.GRCm39 100%[===================>] 769.06M  38.8MB/s    in 20s     \n",
      "\n",
      "2026-01-08 16:32:51 (37.6 MB/s) - ‚Äòdata/reference/mouse_genome.fa.gz‚Äô saved [806418890]\n",
      "\n",
      "--2026-01-08 16:32:51--  ftp://ftp.ensembl.org/pub/release-115/gtf/mus_musculus/Mus_musculus.GRCm39.115.gtf.gz\n",
      "           => ‚Äòdata/reference/mouse_annotation.gtf.gz‚Äô\n",
      "Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.193.169\n",
      "Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.193.169|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /pub/release-115/gtf/mus_musculus ... done.\n",
      "==> SIZE Mus_musculus.GRCm39.115.gtf.gz ... 41681142\n",
      "==> PASV ... done.    ==> RETR Mus_musculus.GRCm39.115.gtf.gz ... done.\n",
      "Length: 41681142 (40M) (unauthoritative)\n",
      "\n",
      "Mus_musculus.GRCm39 100%[===================>]  39.75M  25.9MB/s    in 1.5s    \n",
      "\n",
      "2026-01-08 16:32:54 (25.9 MB/s) - ‚Äòdata/reference/mouse_annotation.gtf.gz‚Äô saved [41681142]\n",
      "\n",
      "--2026-01-08 16:32:54--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.27_GRCm39/GCF_000001635.27_GRCm39_feature_table.txt.gz\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.13, 130.14.250.31, 130.14.250.7, ...\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.13|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 6572982 (6.3M) [application/x-gzip]\n",
      "Saving to: ‚Äòdata/reference/mouse_feature_table.txt.gz‚Äô\n",
      "\n",
      "data/reference/mous 100%[===================>]   6.27M  7.67MB/s    in 0.8s    \n",
      "\n",
      "2026-01-08 16:32:55 (7.67 MB/s) - ‚Äòdata/reference/mouse_feature_table.txt.gz‚Äô saved [6572982/6572982]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget ftp://ftp.ensembl.org/pub/release-115/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.primary_assembly.fa.gz -O data/reference/mouse_genome.fa.gz\n",
    "!wget ftp://ftp.ensembl.org/pub/release-115/gtf/mus_musculus/Mus_musculus.GRCm39.115.gtf.gz -O data/reference/mouse_annotation.gtf.gz\n",
    "!wget -O data/reference/mouse_feature_table.txt.gz https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.27_GRCm39/GCF_000001635.27_GRCm39_feature_table.txt.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "!gunzip -f data/reference/mouse_genome.fa.gz \n",
    "!gunzip -f data/reference/mouse_annotation.gtf.gz\n",
    "!gunzip -f data/reference/mouse_feature_table.txt.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 3.4: Copy data file for Trimmomatic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One of the functions of Trimmomatic is to trim adapter sequences unique to each sequencing platform. These adapter sequences are typically located within the trimmomatic installation directory in a folder called adapters.\n",
    "\n",
    "Directories of packages within conda installations can be confusing, so in the case of using conda with Trimmomatic, it may be easier to simply download or create a file with the relevant adapter sequences and store it in an easy to find directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cp TruSeq3-SE.fa data/trimmed/."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STEP 4: Run FastQC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "FastQC is an invaluable tool that allows you to evaluate whether there are problems with a set of reads. For example, it will provide a report of whether there is any bias in the sequence composition of the reads.\n",
    "\n",
    "The below code may take a while to run. To make it run faster we can use threads to speed up the process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR1810176_1.fastq.gz\n",
      "Started analysis of SRR1810188_1.fastq.gz\n",
      "Started analysis of SRR1810171_1.fastq.gz\n",
      "application/gzip\n",
      "Started analysis of SRR1810156_1.fastq.gz\n",
      "Started analysis of SRR1810163_1.fastq.gz\n",
      "Started analysis of SRR1810158_1.fastq.gz\n",
      "Started analysis of SRR1810162_1.fastq.gz\n",
      "Started analysis of SRR1810167_1.fastq.gz\n",
      "Approx 5% complete for SRR1810156_1.fastq.gz\n",
      "Approx 5% complete for SRR1810171_1.fastq.gz\n",
      "Approx 5% complete for SRR1810162_1.fastq.gz\n",
      "Approx 5% complete for SRR1810158_1.fastq.gz\n",
      "Approx 5% complete for SRR1810163_1.fastq.gz\n",
      "Approx 5% complete for SRR1810188_1.fastq.gz\n",
      "Approx 5% complete for SRR1810167_1.fastq.gz\n",
      "Approx 5% complete for SRR1810176_1.fastq.gz\n",
      "Approx 10% complete for SRR1810156_1.fastq.gz\n",
      "Approx 10% complete for SRR1810171_1.fastq.gz\n",
      "Approx 10% complete for SRR1810162_1.fastq.gz\n",
      "Approx 10% complete for SRR1810158_1.fastq.gz\n",
      "Approx 10% complete for SRR1810188_1.fastq.gz\n",
      "Approx 10% complete for SRR1810163_1.fastq.gz\n",
      "Approx 10% complete for SRR1810167_1.fastq.gz\n",
      "Approx 10% complete for SRR1810176_1.fastq.gz\n",
      "Approx 15% complete for SRR1810156_1.fastq.gz\n",
      "Approx 15% complete for SRR1810171_1.fastq.gz\n",
      "Approx 15% complete for SRR1810162_1.fastq.gz\n",
      "Approx 20% complete for SRR1810156_1.fastq.gz\n",
      "Approx 15% complete for SRR1810158_1.fastq.gz\n",
      "Approx 15% complete for SRR1810188_1.fastq.gz\n",
      "Approx 15% complete for SRR1810163_1.fastq.gz\n",
      "Approx 15% complete for SRR1810167_1.fastq.gz\n",
      "Approx 15% complete for SRR1810176_1.fastq.gz\n",
      "Approx 20% complete for SRR1810171_1.fastq.gz\n",
      "Approx 25% complete for SRR1810156_1.fastq.gz\n",
      "Approx 20% complete for SRR1810162_1.fastq.gz\n",
      "Approx 20% complete for SRR1810158_1.fastq.gz\n",
      "Approx 20% complete for SRR1810188_1.fastq.gz\n",
      "Approx 20% complete for SRR1810163_1.fastq.gz\n",
      "Approx 20% complete for SRR1810167_1.fastq.gz\n",
      "Approx 25% complete for SRR1810171_1.fastq.gz\n",
      "Approx 20% complete for SRR1810176_1.fastq.gz\n",
      "Approx 30% complete for SRR1810156_1.fastq.gz\n",
      "Approx 25% complete for SRR1810162_1.fastq.gz\n",
      "Approx 25% complete for SRR1810158_1.fastq.gz\n",
      "Approx 25% complete for SRR1810188_1.fastq.gz\n",
      "Approx 25% complete for SRR1810163_1.fastq.gz\n",
      "Approx 30% complete for SRR1810171_1.fastq.gz\n",
      "Approx 35% complete for SRR1810156_1.fastq.gz\n",
      "Approx 25% complete for SRR1810167_1.fastq.gz\n",
      "Approx 25% complete for SRR1810176_1.fastq.gz\n",
      "Approx 30% complete for SRR1810162_1.fastq.gz\n",
      "Approx 40% complete for SRR1810156_1.fastq.gz\n",
      "Approx 30% complete for SRR1810158_1.fastq.gz\n",
      "Approx 35% complete for SRR1810171_1.fastq.gz\n",
      "Approx 30% complete for SRR1810188_1.fastq.gz\n",
      "Approx 30% complete for SRR1810163_1.fastq.gz\n",
      "Approx 30% complete for SRR1810167_1.fastq.gz\n",
      "Approx 30% complete for SRR1810176_1.fastq.gz\n",
      "Approx 45% complete for SRR1810156_1.fastq.gz\n",
      "Approx 35% complete for SRR1810162_1.fastq.gz\n",
      "Approx 40% complete for SRR1810171_1.fastq.gz\n",
      "Approx 35% complete for SRR1810158_1.fastq.gz\n",
      "Approx 35% complete for SRR1810188_1.fastq.gz\n",
      "Approx 35% complete for SRR1810163_1.fastq.gz\n",
      "Approx 50% complete for SRR1810156_1.fastq.gz\n",
      "Approx 35% complete for SRR1810167_1.fastq.gz\n",
      "Approx 35% complete for SRR1810176_1.fastq.gz\n",
      "Approx 45% complete for SRR1810171_1.fastq.gz\n",
      "Approx 40% complete for SRR1810162_1.fastq.gz\n",
      "Approx 40% complete for SRR1810158_1.fastq.gz\n",
      "Approx 40% complete for SRR1810188_1.fastq.gz\n",
      "Approx 40% complete for SRR1810163_1.fastq.gz\n",
      "Approx 55% complete for SRR1810156_1.fastq.gz\n",
      "Approx 40% complete for SRR1810167_1.fastq.gz\n",
      "Approx 50% complete for SRR1810171_1.fastq.gz\n",
      "Approx 40% complete for SRR1810176_1.fastq.gz\n",
      "Approx 45% complete for SRR1810162_1.fastq.gz\n",
      "Approx 60% complete for SRR1810156_1.fastq.gz\n",
      "Approx 45% complete for SRR1810158_1.fastq.gz\n",
      "Approx 45% complete for SRR1810188_1.fastq.gz\n",
      "Approx 45% complete for SRR1810163_1.fastq.gz\n",
      "Approx 55% complete for SRR1810171_1.fastq.gz\n",
      "Approx 45% complete for SRR1810167_1.fastq.gz\n",
      "Approx 65% complete for SRR1810156_1.fastq.gz\n",
      "Approx 50% complete for SRR1810162_1.fastq.gz\n",
      "Approx 45% complete for SRR1810176_1.fastq.gz\n",
      "Approx 50% complete for SRR1810158_1.fastq.gz\n",
      "Approx 50% complete for SRR1810188_1.fastq.gz\n",
      "Approx 50% complete for SRR1810163_1.fastq.gz\n",
      "Approx 60% complete for SRR1810171_1.fastq.gz\n",
      "Approx 70% complete for SRR1810156_1.fastq.gz\n",
      "Approx 50% complete for SRR1810167_1.fastq.gz\n",
      "Approx 55% complete for SRR1810162_1.fastq.gz\n",
      "Approx 55% complete for SRR1810158_1.fastq.gz\n",
      "Approx 50% complete for SRR1810176_1.fastq.gz\n",
      "Approx 55% complete for SRR1810188_1.fastq.gz\n",
      "Approx 75% complete for SRR1810156_1.fastq.gz\n",
      "Approx 55% complete for SRR1810163_1.fastq.gz\n",
      "Approx 65% complete for SRR1810171_1.fastq.gz\n",
      "Approx 55% complete for SRR1810167_1.fastq.gz\n",
      "Approx 60% complete for SRR1810162_1.fastq.gz\n",
      "Approx 80% complete for SRR1810156_1.fastq.gz\n",
      "Approx 60% complete for SRR1810158_1.fastq.gz\n",
      "Approx 70% complete for SRR1810171_1.fastq.gz\n",
      "Approx 55% complete for SRR1810176_1.fastq.gz\n",
      "Approx 60% complete for SRR1810188_1.fastq.gz\n",
      "Approx 60% complete for SRR1810163_1.fastq.gz\n",
      "Approx 85% complete for SRR1810156_1.fastq.gz\n",
      "Approx 60% complete for SRR1810167_1.fastq.gz\n",
      "Approx 65% complete for SRR1810162_1.fastq.gz\n",
      "Approx 75% complete for SRR1810171_1.fastq.gz\n",
      "Approx 65% complete for SRR1810158_1.fastq.gz\n",
      "Approx 65% complete for SRR1810188_1.fastq.gz\n",
      "Approx 60% complete for SRR1810176_1.fastq.gz\n",
      "Approx 65% complete for SRR1810163_1.fastq.gz\n",
      "Approx 90% complete for SRR1810156_1.fastq.gz\n",
      "Approx 70% complete for SRR1810162_1.fastq.gz\n",
      "Approx 80% complete for SRR1810171_1.fastq.gz\n",
      "Approx 65% complete for SRR1810167_1.fastq.gz\n",
      "Approx 70% complete for SRR1810158_1.fastq.gz\n",
      "Approx 70% complete for SRR1810188_1.fastq.gz\n",
      "Approx 95% complete for SRR1810156_1.fastq.gz\n",
      "Approx 70% complete for SRR1810163_1.fastq.gz\n",
      "Approx 65% complete for SRR1810176_1.fastq.gz\n",
      "Approx 85% complete for SRR1810171_1.fastq.gz\n",
      "Approx 75% complete for SRR1810162_1.fastq.gz\n",
      "Approx 70% complete for SRR1810167_1.fastq.gz\n",
      "Analysis complete for SRR1810156_1.fastq.gz\n",
      "Approx 75% complete for SRR1810158_1.fastq.gz\n",
      "Approx 75% complete for SRR1810188_1.fastq.gz\n",
      "Approx 75% complete for SRR1810163_1.fastq.gz\n",
      "Approx 70% complete for SRR1810176_1.fastq.gz\n",
      "Approx 90% complete for SRR1810171_1.fastq.gz\n",
      "Approx 80% complete for SRR1810162_1.fastq.gz\n",
      "Approx 75% complete for SRR1810167_1.fastq.gz\n",
      "Approx 80% complete for SRR1810158_1.fastq.gz\n",
      "Approx 80% complete for SRR1810188_1.fastq.gz\n",
      "Approx 80% complete for SRR1810163_1.fastq.gz\n",
      "Approx 95% complete for SRR1810171_1.fastq.gz\n",
      "Approx 75% complete for SRR1810176_1.fastq.gz\n",
      "Approx 85% complete for SRR1810162_1.fastq.gz\n",
      "Approx 85% complete for SRR1810158_1.fastq.gz\n",
      "Approx 80% complete for SRR1810167_1.fastq.gz\n",
      "Approx 85% complete for SRR1810188_1.fastq.gz\n",
      "Analysis complete for SRR1810171_1.fastq.gz\n",
      "Approx 85% complete for SRR1810163_1.fastq.gz\n",
      "Approx 80% complete for SRR1810176_1.fastq.gz\n",
      "Approx 90% complete for SRR1810162_1.fastq.gz\n",
      "Approx 90% complete for SRR1810158_1.fastq.gz\n",
      "Approx 85% complete for SRR1810167_1.fastq.gz\n",
      "Approx 90% complete for SRR1810188_1.fastq.gz\n",
      "Approx 90% complete for SRR1810163_1.fastq.gz\n",
      "Approx 85% complete for SRR1810176_1.fastq.gz\n",
      "Approx 95% complete for SRR1810162_1.fastq.gz\n",
      "Approx 95% complete for SRR1810158_1.fastq.gz\n",
      "Approx 90% complete for SRR1810167_1.fastq.gz\n",
      "Approx 95% complete for SRR1810188_1.fastq.gz\n",
      "Approx 95% complete for SRR1810163_1.fastq.gz\n",
      "Analysis complete for SRR1810162_1.fastq.gz\n",
      "Approx 90% complete for SRR1810176_1.fastq.gz\n",
      "Analysis complete for SRR1810158_1.fastq.gz\n",
      "Analysis complete for SRR1810188_1.fastq.gz\n",
      "Approx 95% complete for SRR1810167_1.fastq.gz\n",
      "Analysis complete for SRR1810163_1.fastq.gz\n",
      "Approx 95% complete for SRR1810176_1.fastq.gz\n",
      "Analysis complete for SRR1810167_1.fastq.gz\n",
      "Analysis complete for SRR1810176_1.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "# Run fastqc for forward reads in parallel\n",
    "!cat accs_GSE66040_set1.txt | xargs -P $THREADS -I {} fastqc \"data/raw_fastq/{}.fastq.gz\" -o data/fastqc/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "FastQC will output the results in HTML format, as below, for all forward and reverse reads.\n",
    "\n",
    "Make sure that you change the SRR run ID in name of the HTML file to be from one samples you analyzed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"800\"\n",
       "            height=\"600\"\n",
       "            src=\"data/fastqc/SRR1810156_1_fastqc.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f6245c6aa40>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "IFrame(src='data/fastqc/SRR1810156_fastqc.html', width=800, height=600)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although it is best practice to look over the quality reports individually, tools like MultiQC allow one to quickly look at a combined summary of the quality reports of all fastq files.\n",
    "\n",
    "For instance, the below table shows all warnings, passes, or failures, from each FastQC report. There are other summaries created as well by MultiQC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[91m///\u001b[0m \u001b]8;id=549824;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ üîç \u001b[2mv1.33\u001b[0m\n",
      "\n",
      "\u001b[34m       file_search\u001b[0m | Search path: /home/jupyter/mouse-iav-infection/data/fastqc\n",
      "\u001b[2K         \u001b[34msearching\u001b[0m | \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m16/16\u001b[0m  mdata/fastqc/SRR1810171_1_fastqc.html\u001b[0m\n",
      "\u001b[?25h\u001b[34m            fastqc\u001b[0m | Found 8 reports\n",
      "\u001b[34m     write_results\u001b[0m | Data        : multiqc_data\n",
      "\u001b[34m     write_results\u001b[0m | Report      : multiqc_report.html\n",
      "\u001b[34m           multiqc\u001b[0m | MultiQC complete\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Filename</th>\n",
       "      <th>File type</th>\n",
       "      <th>Encoding</th>\n",
       "      <th>Total Sequences</th>\n",
       "      <th>Total Bases</th>\n",
       "      <th>Sequences flagged as poor quality</th>\n",
       "      <th>Sequence length</th>\n",
       "      <th>%GC</th>\n",
       "      <th>total_deduplicated_percentage</th>\n",
       "      <th>...</th>\n",
       "      <th>basic_statistics</th>\n",
       "      <th>per_base_sequence_quality</th>\n",
       "      <th>per_sequence_quality_scores</th>\n",
       "      <th>per_base_sequence_content</th>\n",
       "      <th>per_sequence_gc_content</th>\n",
       "      <th>per_base_n_content</th>\n",
       "      <th>sequence_length_distribution</th>\n",
       "      <th>sequence_duplication_levels</th>\n",
       "      <th>overrepresented_sequences</th>\n",
       "      <th>adapter_content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR1810156_1</td>\n",
       "      <td>SRR1810156_1.fastq.gz</td>\n",
       "      <td>Colorspace converted to bases</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>17088302.0</td>\n",
       "      <td>1.1 Gbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25-75</td>\n",
       "      <td>49.0</td>\n",
       "      <td>71.038681</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR1810158_1</td>\n",
       "      <td>SRR1810158_1.fastq.gz</td>\n",
       "      <td>Colorspace converted to bases</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>23414444.0</td>\n",
       "      <td>1.5 Gbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25-75</td>\n",
       "      <td>48.0</td>\n",
       "      <td>62.620503</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>warn</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR1810162_1</td>\n",
       "      <td>SRR1810162_1.fastq.gz</td>\n",
       "      <td>Colorspace converted to bases</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>22942035.0</td>\n",
       "      <td>1.5 Gbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25-75</td>\n",
       "      <td>48.0</td>\n",
       "      <td>68.809351</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>warn</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR1810163_1</td>\n",
       "      <td>SRR1810163_1.fastq.gz</td>\n",
       "      <td>Colorspace converted to bases</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>23907580.0</td>\n",
       "      <td>1.5 Gbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25-75</td>\n",
       "      <td>50.0</td>\n",
       "      <td>69.683134</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>warn</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR1810167_1</td>\n",
       "      <td>SRR1810167_1.fastq.gz</td>\n",
       "      <td>Colorspace converted to bases</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>25525585.0</td>\n",
       "      <td>1.6 Gbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25-75</td>\n",
       "      <td>48.0</td>\n",
       "      <td>66.580729</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>warn</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR1810171_1</td>\n",
       "      <td>SRR1810171_1.fastq.gz</td>\n",
       "      <td>Colorspace converted to bases</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>20661559.0</td>\n",
       "      <td>1.4 Gbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25-75</td>\n",
       "      <td>48.0</td>\n",
       "      <td>65.132883</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>warn</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRR1810176_1</td>\n",
       "      <td>SRR1810176_1.fastq.gz</td>\n",
       "      <td>Colorspace converted to bases</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>26138852.0</td>\n",
       "      <td>1.7 Gbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25-75</td>\n",
       "      <td>50.0</td>\n",
       "      <td>69.357255</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>warn</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRR1810188_1</td>\n",
       "      <td>SRR1810188_1.fastq.gz</td>\n",
       "      <td>Colorspace converted to bases</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>24241421.0</td>\n",
       "      <td>1.5 Gbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25-75</td>\n",
       "      <td>51.0</td>\n",
       "      <td>73.656359</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows √ó 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Sample               Filename                      File type  \\\n",
       "0  SRR1810156_1  SRR1810156_1.fastq.gz  Colorspace converted to bases   \n",
       "1  SRR1810158_1  SRR1810158_1.fastq.gz  Colorspace converted to bases   \n",
       "2  SRR1810162_1  SRR1810162_1.fastq.gz  Colorspace converted to bases   \n",
       "3  SRR1810163_1  SRR1810163_1.fastq.gz  Colorspace converted to bases   \n",
       "4  SRR1810167_1  SRR1810167_1.fastq.gz  Colorspace converted to bases   \n",
       "5  SRR1810171_1  SRR1810171_1.fastq.gz  Colorspace converted to bases   \n",
       "6  SRR1810176_1  SRR1810176_1.fastq.gz  Colorspace converted to bases   \n",
       "7  SRR1810188_1  SRR1810188_1.fastq.gz  Colorspace converted to bases   \n",
       "\n",
       "                Encoding  Total Sequences Total Bases  \\\n",
       "0  Sanger / Illumina 1.9       17088302.0     1.1 Gbp   \n",
       "1  Sanger / Illumina 1.9       23414444.0     1.5 Gbp   \n",
       "2  Sanger / Illumina 1.9       22942035.0     1.5 Gbp   \n",
       "3  Sanger / Illumina 1.9       23907580.0     1.5 Gbp   \n",
       "4  Sanger / Illumina 1.9       25525585.0     1.6 Gbp   \n",
       "5  Sanger / Illumina 1.9       20661559.0     1.4 Gbp   \n",
       "6  Sanger / Illumina 1.9       26138852.0     1.7 Gbp   \n",
       "7  Sanger / Illumina 1.9       24241421.0     1.5 Gbp   \n",
       "\n",
       "   Sequences flagged as poor quality Sequence length   %GC  \\\n",
       "0                                0.0           25-75  49.0   \n",
       "1                                0.0           25-75  48.0   \n",
       "2                                0.0           25-75  48.0   \n",
       "3                                0.0           25-75  50.0   \n",
       "4                                0.0           25-75  48.0   \n",
       "5                                0.0           25-75  48.0   \n",
       "6                                0.0           25-75  50.0   \n",
       "7                                0.0           25-75  51.0   \n",
       "\n",
       "   total_deduplicated_percentage  ...  basic_statistics  \\\n",
       "0                      71.038681  ...              pass   \n",
       "1                      62.620503  ...              pass   \n",
       "2                      68.809351  ...              pass   \n",
       "3                      69.683134  ...              pass   \n",
       "4                      66.580729  ...              pass   \n",
       "5                      65.132883  ...              pass   \n",
       "6                      69.357255  ...              pass   \n",
       "7                      73.656359  ...              pass   \n",
       "\n",
       "   per_base_sequence_quality per_sequence_quality_scores  \\\n",
       "0                       fail                        pass   \n",
       "1                       fail                        pass   \n",
       "2                       fail                        pass   \n",
       "3                       fail                        pass   \n",
       "4                       fail                        pass   \n",
       "5                       fail                        pass   \n",
       "6                       fail                        pass   \n",
       "7                       fail                        pass   \n",
       "\n",
       "  per_base_sequence_content per_sequence_gc_content per_base_n_content  \\\n",
       "0                      fail                    fail               fail   \n",
       "1                      fail                    fail               fail   \n",
       "2                      fail                    fail               fail   \n",
       "3                      fail                    fail               fail   \n",
       "4                      fail                    fail               fail   \n",
       "5                      fail                    fail               fail   \n",
       "6                      fail                    fail               fail   \n",
       "7                      fail                    fail               fail   \n",
       "\n",
       "  sequence_length_distribution sequence_duplication_levels  \\\n",
       "0                         warn                        pass   \n",
       "1                         warn                        warn   \n",
       "2                         warn                        warn   \n",
       "3                         warn                        warn   \n",
       "4                         warn                        warn   \n",
       "5                         warn                        warn   \n",
       "6                         warn                        warn   \n",
       "7                         warn                        pass   \n",
       "\n",
       "  overrepresented_sequences adapter_content  \n",
       "0                      fail            pass  \n",
       "1                      fail            pass  \n",
       "2                      fail            pass  \n",
       "3                      fail            pass  \n",
       "4                      fail            pass  \n",
       "5                      fail            pass  \n",
       "6                      fail            pass  \n",
       "7                      fail            pass  \n",
       "\n",
       "[8 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "!multiqc -f data/fastqc/\n",
    "\n",
    "import pandas as pd\n",
    "dframe = pd.read_csv(\"./multiqc_data/multiqc_fastqc.txt\", sep='\\t')\n",
    "display(dframe)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STEP 5: Run Trimmomatic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trimmomatic will trim off any adapter sequences or low quality sequence it detects in the FASTQ files.\n",
    "\n",
    "Using piping and our original list, it is possible to queue up a batch run of Trimmomatic for all our files, note that this is a different way to run a loop compared with what we did before.\n",
    "\n",
    "The below code may take approximately 30 minutes to run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TrimmomaticSE: Started with arguments:\n",
      " -threads 31 data/raw_fastq/SRR1810156.fastq data/trimmed/SRR1810156_trimmed.fastq ILLUMINACLIP:data/trimmed/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
      "ILLUMINACLIP: Using adapter file from current working directory: /home/jupyter/mouse-iav-infection/data/trimmed/TruSeq3-SE.fa\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA'\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 17088302 Surviving: 10300142 (60.28%) Dropped: 6788160 (39.72%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " -threads 31 data/raw_fastq/SRR1810158.fastq data/trimmed/SRR1810158_trimmed.fastq ILLUMINACLIP:data/trimmed/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
      "ILLUMINACLIP: Using adapter file from current working directory: /home/jupyter/mouse-iav-infection/data/trimmed/TruSeq3-SE.fa\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA'\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 23414444 Surviving: 14136270 (60.37%) Dropped: 9278174 (39.63%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " -threads 31 data/raw_fastq/SRR1810162.fastq data/trimmed/SRR1810162_trimmed.fastq ILLUMINACLIP:data/trimmed/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
      "ILLUMINACLIP: Using adapter file from current working directory: /home/jupyter/mouse-iav-infection/data/trimmed/TruSeq3-SE.fa\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA'\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 22942035 Surviving: 13946170 (60.79%) Dropped: 8995865 (39.21%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " -threads 31 data/raw_fastq/SRR1810163.fastq data/trimmed/SRR1810163_trimmed.fastq ILLUMINACLIP:data/trimmed/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
      "ILLUMINACLIP: Using adapter file from current working directory: /home/jupyter/mouse-iav-infection/data/trimmed/TruSeq3-SE.fa\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA'\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 23907580 Surviving: 13855461 (57.95%) Dropped: 10052119 (42.05%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " -threads 31 data/raw_fastq/SRR1810167.fastq data/trimmed/SRR1810167_trimmed.fastq ILLUMINACLIP:data/trimmed/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
      "ILLUMINACLIP: Using adapter file from current working directory: /home/jupyter/mouse-iav-infection/data/trimmed/TruSeq3-SE.fa\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA'\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 25525585 Surviving: 15601222 (61.12%) Dropped: 9924363 (38.88%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " -threads 31 data/raw_fastq/SRR1810171.fastq data/trimmed/SRR1810171_trimmed.fastq ILLUMINACLIP:data/trimmed/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
      "ILLUMINACLIP: Using adapter file from current working directory: /home/jupyter/mouse-iav-infection/data/trimmed/TruSeq3-SE.fa\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA'\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 20661559 Surviving: 12323317 (59.64%) Dropped: 8338242 (40.36%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " -threads 31 data/raw_fastq/SRR1810176.fastq data/trimmed/SRR1810176_trimmed.fastq ILLUMINACLIP:data/trimmed/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
      "ILLUMINACLIP: Using adapter file from current working directory: /home/jupyter/mouse-iav-infection/data/trimmed/TruSeq3-SE.fa\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA'\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 26138852 Surviving: 15563160 (59.54%) Dropped: 10575692 (40.46%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " -threads 31 data/raw_fastq/SRR1810188.fastq data/trimmed/SRR1810188_trimmed.fastq ILLUMINACLIP:data/trimmed/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
      "ILLUMINACLIP: Using adapter file from current working directory: /home/jupyter/mouse-iav-infection/data/trimmed/TruSeq3-SE.fa\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA'\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 24241421 Surviving: 13409629 (55.32%) Dropped: 10831792 (44.68%)\n",
      "TrimmomaticSE: Completed successfully\n"
     ]
    }
   ],
   "source": [
    "!cat accs_GSE66040_set1.txt | xargs -I {} \\\n",
    "trimmomatic SE -threads $THREADS \\\n",
    "'data/raw_fastq/{}.fastq' \\\n",
    "'data/trimmed/{}_trimmed.fastq' \\\n",
    "ILLUMINACLIP:data/trimmed/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STEP 6: Run FastQC\n",
    "It's best practice to run FastQC after trimming. However, you may decide to run FastQC only once, before or after trimming.\n",
    "\n",
    "We will proceed with only the forward reads -- this is because, looking at Trimmomatic, there were very few 'orphaned' reads. That is to say, most forward and reverse reads were successfully paired together. Because we are just trying to map to a reference genome, the read lengths of the forward reads alone, in this case, around 60 million basepairs, should be sufficient.\n",
    "\n",
    "The below code may take around 15-20 minutes to run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null\n",
      "null\n",
      "null\n",
      "null\n",
      "null\n",
      "null\n",
      "null\n",
      "Started analysis of SRR1810163_trimmed.fastq\n",
      "Started analysis of SRR1810158_trimmed.fastq\n",
      "null\n",
      "Started analysis of SRR1810167_trimmed.fastq\n",
      "Started analysis of SRR1810171_trimmed.fastq\n",
      "Started analysis of SRR1810188_trimmed.fastq\n",
      "Started analysis of SRR1810156_trimmed.fastq\n",
      "Started analysis of SRR1810162_trimmed.fastq\n",
      "Started analysis of SRR1810176_trimmed.fastq\n",
      "Approx 5% complete for SRR1810156_trimmed.fastq\n",
      "Approx 5% complete for SRR1810188_trimmed.fastq\n",
      "Approx 5% complete for SRR1810171_trimmed.fastq\n",
      "Approx 5% complete for SRR1810163_trimmed.fastq\n",
      "Approx 5% complete for SRR1810158_trimmed.fastq\n",
      "Approx 5% complete for SRR1810167_trimmed.fastq\n",
      "Approx 5% complete for SRR1810162_trimmed.fastq\n",
      "Approx 5% complete for SRR1810176_trimmed.fastq\n",
      "Approx 10% complete for SRR1810156_trimmed.fastq\n",
      "Approx 10% complete for SRR1810188_trimmed.fastq\n",
      "Approx 10% complete for SRR1810171_trimmed.fastq\n",
      "Approx 10% complete for SRR1810163_trimmed.fastq\n",
      "Approx 10% complete for SRR1810158_trimmed.fastq\n",
      "Approx 10% complete for SRR1810167_trimmed.fastq\n",
      "Approx 15% complete for SRR1810156_trimmed.fastq\n",
      "Approx 10% complete for SRR1810162_trimmed.fastq\n",
      "Approx 10% complete for SRR1810176_trimmed.fastq\n",
      "Approx 15% complete for SRR1810188_trimmed.fastq\n",
      "Approx 15% complete for SRR1810171_trimmed.fastq\n",
      "Approx 15% complete for SRR1810163_trimmed.fastq\n",
      "Approx 20% complete for SRR1810156_trimmed.fastq\n",
      "Approx 15% complete for SRR1810158_trimmed.fastq\n",
      "Approx 15% complete for SRR1810162_trimmed.fastq\n",
      "Approx 15% complete for SRR1810167_trimmed.fastq\n",
      "Approx 20% complete for SRR1810188_trimmed.fastq\n",
      "Approx 20% complete for SRR1810171_trimmed.fastq\n",
      "Approx 15% complete for SRR1810176_trimmed.fastq\n",
      "Approx 25% complete for SRR1810156_trimmed.fastq\n",
      "Approx 20% complete for SRR1810163_trimmed.fastq\n",
      "Approx 20% complete for SRR1810158_trimmed.fastq\n",
      "Approx 20% complete for SRR1810162_trimmed.fastq\n",
      "Approx 20% complete for SRR1810167_trimmed.fastq\n",
      "Approx 30% complete for SRR1810156_trimmed.fastq\n",
      "Approx 25% complete for SRR1810171_trimmed.fastq\n",
      "Approx 25% complete for SRR1810188_trimmed.fastq\n",
      "Approx 20% complete for SRR1810176_trimmed.fastq\n",
      "Approx 25% complete for SRR1810163_trimmed.fastq\n",
      "Approx 25% complete for SRR1810158_trimmed.fastq\n",
      "Approx 25% complete for SRR1810162_trimmed.fastq\n",
      "Approx 35% complete for SRR1810156_trimmed.fastq\n",
      "Approx 30% complete for SRR1810171_trimmed.fastq\n",
      "Approx 30% complete for SRR1810188_trimmed.fastq\n",
      "Approx 25% complete for SRR1810167_trimmed.fastq\n",
      "Approx 25% complete for SRR1810176_trimmed.fastq\n",
      "Approx 30% complete for SRR1810163_trimmed.fastq\n",
      "Approx 40% complete for SRR1810156_trimmed.fastq\n",
      "Approx 30% complete for SRR1810158_trimmed.fastq\n",
      "Approx 30% complete for SRR1810162_trimmed.fastq\n",
      "Approx 35% complete for SRR1810171_trimmed.fastq\n",
      "Approx 35% complete for SRR1810188_trimmed.fastq\n",
      "Approx 30% complete for SRR1810167_trimmed.fastq\n",
      "Approx 45% complete for SRR1810156_trimmed.fastq\n",
      "Approx 35% complete for SRR1810163_trimmed.fastq\n",
      "Approx 30% complete for SRR1810176_trimmed.fastq\n",
      "Approx 35% complete for SRR1810158_trimmed.fastq\n",
      "Approx 35% complete for SRR1810162_trimmed.fastq\n",
      "Approx 40% complete for SRR1810188_trimmed.fastq\n",
      "Approx 40% complete for SRR1810171_trimmed.fastq\n",
      "Approx 50% complete for SRR1810156_trimmed.fastq\n",
      "Approx 35% complete for SRR1810167_trimmed.fastq\n",
      "Approx 40% complete for SRR1810163_trimmed.fastq\n",
      "Approx 35% complete for SRR1810176_trimmed.fastq\n",
      "Approx 40% complete for SRR1810158_trimmed.fastq\n",
      "Approx 40% complete for SRR1810162_trimmed.fastq\n",
      "Approx 45% complete for SRR1810188_trimmed.fastq\n",
      "Approx 45% complete for SRR1810171_trimmed.fastq\n",
      "Approx 55% complete for SRR1810156_trimmed.fastq\n",
      "Approx 40% complete for SRR1810167_trimmed.fastq\n",
      "Approx 45% complete for SRR1810163_trimmed.fastq\n",
      "Approx 60% complete for SRR1810156_trimmed.fastq\n",
      "Approx 50% complete for SRR1810188_trimmed.fastq\n",
      "Approx 45% complete for SRR1810158_trimmed.fastq\n",
      "Approx 50% complete for SRR1810171_trimmed.fastq\n",
      "Approx 45% complete for SRR1810162_trimmed.fastq\n",
      "Approx 40% complete for SRR1810176_trimmed.fastq\n",
      "Approx 65% complete for SRR1810156_trimmed.fastq\n",
      "Approx 50% complete for SRR1810163_trimmed.fastq\n",
      "Approx 45% complete for SRR1810167_trimmed.fastq\n",
      "Approx 55% complete for SRR1810188_trimmed.fastq\n",
      "Approx 55% complete for SRR1810171_trimmed.fastq\n",
      "Approx 50% complete for SRR1810158_trimmed.fastq\n",
      "Approx 50% complete for SRR1810162_trimmed.fastq\n",
      "Approx 45% complete for SRR1810176_trimmed.fastq\n",
      "Approx 70% complete for SRR1810156_trimmed.fastq\n",
      "Approx 55% complete for SRR1810163_trimmed.fastq\n",
      "Approx 60% complete for SRR1810188_trimmed.fastq\n",
      "Approx 50% complete for SRR1810167_trimmed.fastq\n",
      "Approx 60% complete for SRR1810171_trimmed.fastq\n",
      "Approx 75% complete for SRR1810156_trimmed.fastq\n",
      "Approx 55% complete for SRR1810158_trimmed.fastq\n",
      "Approx 55% complete for SRR1810162_trimmed.fastq\n",
      "Approx 50% complete for SRR1810176_trimmed.fastq\n",
      "Approx 60% complete for SRR1810163_trimmed.fastq\n",
      "Approx 65% complete for SRR1810188_trimmed.fastq\n",
      "Approx 80% complete for SRR1810156_trimmed.fastq\n",
      "Approx 65% complete for SRR1810171_trimmed.fastq\n",
      "Approx 55% complete for SRR1810167_trimmed.fastq\n",
      "Approx 60% complete for SRR1810162_trimmed.fastq\n",
      "Approx 60% complete for SRR1810158_trimmed.fastq\n",
      "Approx 55% complete for SRR1810176_trimmed.fastq\n",
      "Approx 85% complete for SRR1810156_trimmed.fastq\n",
      "Approx 70% complete for SRR1810188_trimmed.fastq\n",
      "Approx 65% complete for SRR1810163_trimmed.fastq\n",
      "Approx 70% complete for SRR1810171_trimmed.fastq\n",
      "Approx 60% complete for SRR1810167_trimmed.fastq\n",
      "Approx 65% complete for SRR1810162_trimmed.fastq\n",
      "Approx 65% complete for SRR1810158_trimmed.fastq\n",
      "Approx 90% complete for SRR1810156_trimmed.fastq\n",
      "Approx 75% complete for SRR1810188_trimmed.fastq\n",
      "Approx 60% complete for SRR1810176_trimmed.fastq\n",
      "Approx 70% complete for SRR1810163_trimmed.fastq\n",
      "Approx 75% complete for SRR1810171_trimmed.fastq\n",
      "Approx 95% complete for SRR1810156_trimmed.fastq\n",
      "Approx 70% complete for SRR1810162_trimmed.fastq\n",
      "Approx 70% complete for SRR1810158_trimmed.fastq\n",
      "Approx 65% complete for SRR1810167_trimmed.fastq\n",
      "Approx 80% complete for SRR1810188_trimmed.fastq\n",
      "Approx 80% complete for SRR1810171_trimmed.fastq\n",
      "Approx 75% complete for SRR1810163_trimmed.fastq\n",
      "Approx 65% complete for SRR1810176_trimmed.fastq\n",
      "Analysis complete for SRR1810156_trimmed.fastq\n",
      "Approx 75% complete for SRR1810162_trimmed.fastq\n",
      "Approx 75% complete for SRR1810158_trimmed.fastq\n",
      "Approx 70% complete for SRR1810167_trimmed.fastq\n",
      "Approx 85% complete for SRR1810188_trimmed.fastq\n",
      "Approx 85% complete for SRR1810171_trimmed.fastq\n",
      "Approx 80% complete for SRR1810163_trimmed.fastq\n",
      "Approx 70% complete for SRR1810176_trimmed.fastq\n",
      "Approx 80% complete for SRR1810162_trimmed.fastq\n",
      "Approx 80% complete for SRR1810158_trimmed.fastq\n",
      "Approx 90% complete for SRR1810188_trimmed.fastq\n",
      "Approx 75% complete for SRR1810167_trimmed.fastq\n",
      "Approx 90% complete for SRR1810171_trimmed.fastq\n",
      "Approx 85% complete for SRR1810163_trimmed.fastq\n",
      "Approx 75% complete for SRR1810176_trimmed.fastq\n",
      "Approx 85% complete for SRR1810162_trimmed.fastq\n",
      "Approx 95% complete for SRR1810188_trimmed.fastq\n",
      "Approx 85% complete for SRR1810158_trimmed.fastq\n",
      "Approx 95% complete for SRR1810171_trimmed.fastq\n",
      "Approx 80% complete for SRR1810167_trimmed.fastq\n",
      "Approx 90% complete for SRR1810163_trimmed.fastq\n",
      "Analysis complete for SRR1810188_trimmed.fastq\n",
      "Approx 80% complete for SRR1810176_trimmed.fastq\n",
      "Approx 90% complete for SRR1810162_trimmed.fastq\n",
      "Approx 90% complete for SRR1810158_trimmed.fastq\n",
      "Analysis complete for SRR1810171_trimmed.fastq\n",
      "Approx 85% complete for SRR1810167_trimmed.fastq\n",
      "Approx 95% complete for SRR1810163_trimmed.fastq\n",
      "Approx 95% complete for SRR1810162_trimmed.fastq\n",
      "Approx 85% complete for SRR1810176_trimmed.fastq\n",
      "Approx 95% complete for SRR1810158_trimmed.fastq\n",
      "Approx 90% complete for SRR1810167_trimmed.fastq\n",
      "Analysis complete for SRR1810163_trimmed.fastq\n",
      "Analysis complete for SRR1810162_trimmed.fastq\n",
      "Approx 90% complete for SRR1810176_trimmed.fastq\n",
      "Analysis complete for SRR1810158_trimmed.fastq\n",
      "Approx 95% complete for SRR1810167_trimmed.fastq\n",
      "Approx 95% complete for SRR1810176_trimmed.fastq\n",
      "Analysis complete for SRR1810167_trimmed.fastq\n",
      "Analysis complete for SRR1810176_trimmed.fastq\n"
     ]
    }
   ],
   "source": [
    "# Run FastQC\n",
    "!cat accs_GSE66040_set1.txt | xargs -P $THREADS -I {} fastqc data/trimmed/{}_trimmed.fastq -o data/fastqc_trimmed/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STEP 7: Run MultiQC\n",
    "MultiQC reads in the FastQC reports and generates a compiled report for all the analyzed FASTQ files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[91m///\u001b[0m \u001b]8;id=728182;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ üîç \u001b[2mv1.33\u001b[0m\n",
      "\n",
      "\u001b[34m       file_search\u001b[0m | Search path: /home/jupyter/mouse-iav-infection/data/fastqc_trimmed\n",
      "\u001b[2K         \u001b[34msearching\u001b[0m | \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m16/16\u001b[0m  l\u001b[0mm  \n",
      "\u001b[?25h\u001b[34m            fastqc\u001b[0m | Found 8 reports\n",
      "\u001b[34m     write_results\u001b[0m | Data        : data/multiqc_samples/multiqc_data\n",
      "\u001b[34m     write_results\u001b[0m | Report      : data/multiqc_samples/multiqc_report.html\n",
      "\u001b[34m           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "#!multiqc -f data/fastqc_samples/\n",
    "!multiqc -f -o data/multiqc_samples/ data/fastqc_trimmed/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STEP 8: Preparing the Bowtie-Compatible RSEM Reference"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This command prepares a reference genome and annotation files for RNA-Seq analysis using RSEM (RNA-Seq by Expectation-Maximization) and Bowtie. It generates files needed to quantify gene and isoform expression. The rsem-prepare-reference function takes a GTF file with gene annotations (mouse_annotation.gtf) and a FASTA file with the reference genome sequence (mouse_genome.fa). It processes these files to create a reference, saving the output in the mouse_reference directory. The -p $THREADS option sets the number of threads used for parallel processing, speeding up the preparation process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rsem-extract-reference-transcripts data/rsem_reference/mouse_reference 0 data/reference/mouse_annotation.gtf None 0 data/reference/mouse_genome.fa\n",
      "Parsed 200000 lines\n",
      "Parsed 400000 lines\n",
      "Parsed 600000 lines\n",
      "Parsed 800000 lines\n",
      "Parsed 1000000 lines\n",
      "Parsed 1200000 lines\n",
      "Parsed 1400000 lines\n",
      "Parsed 1600000 lines\n",
      "Parsed 1800000 lines\n",
      "Parsed 2000000 lines\n",
      "Parsed 2200000 lines\n",
      "Parsed 2400000 lines\n",
      "Parsing gtf File is done!\n",
      "data/reference/mouse_genome.fa is processed!\n",
      "278396 transcripts are extracted.\n",
      "Extracting sequences is done!\n",
      "Group File is generated!\n",
      "Transcript Information File is generated!\n",
      "Chromosome List File is generated!\n",
      "Extracted Sequences File is generated!\n",
      "\n",
      "rsem-preref data/rsem_reference/mouse_reference.transcripts.fa 1 data/rsem_reference/mouse_reference\n",
      "Refs.makeRefs finished!\n",
      "Refs.saveRefs finished!\n",
      "data/rsem_reference/mouse_reference.idx.fa is generated!\n",
      "data/rsem_reference/mouse_reference.n2g.idx.fa is generated!\n",
      "\n",
      "bowtie-build -f data/rsem_reference/mouse_reference.n2g.idx.fa data/rsem_reference/mouse_reference\n",
      "Settings:\n",
      "  Output files: \"data/rsem_reference/mouse_reference.*.ebwt\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 5 (one in 32)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  data/rsem_reference/mouse_reference.n2g.idx.fa\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:03\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:02\n",
      "bmax according to bmaxDivN setting: 89816324\n",
      "Using parameters --bmax 67362243 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 67362243 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:21\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:03\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:03\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 5.13236e+07 (target: 67362242)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (67362243) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 62235808 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:47\n",
      "Returning block of 62235809 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (67362243) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 62826526 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:49\n",
      "Returning block of 62826527 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (67362243) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 27119605 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:50\n",
      "Returning block of 27119606 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (67362243) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 56296031 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:39\n",
      "Returning block of 56296032 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (67362243) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 37249699 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:09\n",
      "Returning block of 37249700 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (67362243) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 66257517 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:51\n",
      "Returning block of 66257518 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (67362243) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 47280104 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:19\n",
      "Returning block of 47280105 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 95824933\n",
      "fchr[G]: 181614927\n",
      "fchr[T]: 268838952\n",
      "fchr[$]: 359265296\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 116585605 bytes to primary EBWT file: data/rsem_reference/mouse_reference.1.ebwt\n",
      "Wrote 44908168 bytes to secondary EBWT file: data/rsem_reference/mouse_reference.2.ebwt\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 359265296\n",
      "    bwtLen: 359265297\n",
      "    sz: 89816324\n",
      "    bwtSz: 89816325\n",
      "    lineRate: 6\n",
      "    linesPerSide: 1\n",
      "    offRate: 5\n",
      "    offMask: 0xffffffe0\n",
      "    isaRate: -1\n",
      "    isaMask: 0xffffffff\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 11227041\n",
      "    offsSz: 44908164\n",
      "    isaLen: 0\n",
      "    isaSz: 0\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 56\n",
      "    sideBwtLen: 224\n",
      "    numSidePairs: 801932\n",
      "    numSides: 1603864\n",
      "    numLines: 1603864\n",
      "    ebwtTotLen: 102647296\n",
      "    ebwtTotSz: 102647296\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:12:45\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:02\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:02\n",
      "bmax according to bmaxDivN setting: 89816324\n",
      "Using parameters --bmax 67362243 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 67362243 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:19\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:03\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:04\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 4.49082e+07 (target: 67362242)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (67362243) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 15432988 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:26\n",
      "Returning block of 15432989 for bucket 1\n",
      "Getting block 2 of 8\n",
      "  Reserving size (67362243) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 60410099 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:45\n",
      "Returning block of 60410100 for bucket 2\n",
      "Getting block 3 of 8\n",
      "  Reserving size (67362243) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 53934488 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:35\n",
      "Returning block of 53934489 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (67362243) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 63681189 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:48\n",
      "Returning block of 63681190 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (67362243) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 48053978 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:20\n",
      "Returning block of 48053979 for bucket 5\n",
      "Getting block 6 of 8\n",
      "  Reserving size (67362243) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 42529113 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:14\n",
      "Returning block of 42529114 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (67362243) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 46800118 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:19\n",
      "Returning block of 46800119 for bucket 7\n",
      "Getting block 8 of 8\n",
      "  Reserving size (67362243) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 28423316 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:48\n",
      "Returning block of 28423317 for bucket 8\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 95824933\n",
      "fchr[G]: 181614927\n",
      "fchr[T]: 268838952\n",
      "fchr[$]: 359265296\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 116585605 bytes to primary EBWT file: data/rsem_reference/mouse_reference.rev.1.ebwt\n",
      "Wrote 44908168 bytes to secondary EBWT file: data/rsem_reference/mouse_reference.rev.2.ebwt\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 359265296\n",
      "    bwtLen: 359265297\n",
      "    sz: 89816324\n",
      "    bwtSz: 89816325\n",
      "    lineRate: 6\n",
      "    linesPerSide: 1\n",
      "    offRate: 5\n",
      "    offMask: 0xffffffe0\n",
      "    isaRate: -1\n",
      "    isaMask: 0xffffffff\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 11227041\n",
      "    offsSz: 44908164\n",
      "    isaLen: 0\n",
      "    isaSz: 0\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 56\n",
      "    sideBwtLen: 224\n",
      "    numSidePairs: 801932\n",
      "    numSides: 1603864\n",
      "    numLines: 1603864\n",
      "    ebwtTotLen: 102647296\n",
      "    ebwtTotSz: 102647296\n",
      "    reverse: 0\n",
      "Total time for backward call to driver() for mirror index: 00:12:28\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Preparing the reference genome\n",
    "!rsem-prepare-reference --gtf data/reference/mouse_annotation.gtf --bowtie -p $THREADS data/reference/mouse_genome.fa data/rsem_reference/mouse_reference "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STEP 9: Run Bowtie for Alignment, Prepare and Run RSEM for Quantification"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script automates RNA-Seq gene expression quantification using RSEM and Bowtie. It reads SRR accession IDs from accs.txt, saves results in data/rsem_output, and runs rsem-calculate-expression for each ID. It uses paired-end trimmed FASTQ files from data/trimmed/ and a Bowtie-aligned RSEM reference (mouse_reference)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bowtie -q --phred33-quals -n 2 -e 99999999 -l 25 -p 31 -a -m 200 -S data/rsem_reference/mouse_reference data/trimmed/SRR1810156_trimmed.fastq 2> data/rsem_output/SRR1810156.log | samtools view -b -o data/rsem_output/SRR1810156.temp/SRR1810156.bam -\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Ensure you set the path to the RSEM binary\n",
    "# Read the SRR accessions from the file\n",
    "with open('accs_GSE66040_set1.txt', 'r') as f:\n",
    "    srr_accessions = [line.strip() for line in f.readlines()]\n",
    "\n",
    "# Define the output directory\n",
    "output_dir = \"data/rsem_output\"\n",
    "\n",
    "# Loop through each SRR accession and run rsem-calculate-expression\n",
    "for srr in srr_accessions:\n",
    "    !rsem-calculate-expression -p $THREADS \\\n",
    "    data/trimmed/{srr}_trimmed.fastq data/rsem_reference/mouse_reference data/rsem_output/{srr} "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 10: Report the expression of ENSMUSG00000064351 for each file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use `grep` to report the expression in the wild-type sample. The fields in the RSEM `genes.results` file are as follows. The level of expression is reported in the Transcripts Per Million (`TPM`) and number of reads (`NumReads`) fields:  \n",
    "- `Name`\n",
    "- `Length`\n",
    "- `EffectiveLength`\n",
    "- `TPM`\n",
    "- `NumReads`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/rsem_output/SRR1810156.genes.results:ENSMUSG00000064351\tENSMUST00000082402\t1545.00\t1489.02\t63882.63\t9162.76\t5594.64\n",
      "data/rsem_output/SRR1810158.genes.results:ENSMUSG00000064351\tENSMUST00000082402\t1545.00\t1487.86\t109055.75\t10053.06\t7037.51\n",
      "data/rsem_output/SRR1810162.genes.results:ENSMUSG00000064351\tENSMUST00000082402\t1545.00\t1488.44\t102595.31\t12440.12\t6889.57\n",
      "data/rsem_output/SRR1810167.genes.results:ENSMUSG00000064351\tENSMUST00000082402\t1545.00\t1545.00\t0.00\t0.00\t0.00\n",
      "data/rsem_output/SRR1810171.genes.results:ENSMUSG00000064351\tENSMUST00000082402\t1545.00\t1545.00\t0.00\t0.00\t0.00\n",
      "data/rsem_output/SRR1810176.genes.results:ENSMUSG00000064351\tENSMUST00000082402\t1545.00\t1545.00\t0.00\t0.00\t0.00\n",
      "data/rsem_output/SRR1810188.genes.results:ENSMUSG00000064351\tENSMUST00000082402\t1545.00\t1545.00\t0.00\t0.00\t0.00\n"
     ]
    }
   ],
   "source": [
    "!grep 'ENSMUSG00000064351' data/rsem_output/*.genes.results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusion\n",
    "\n",
    "In this tutorial, we covered the following key concepts and workflow steps:\n",
    "\n",
    "- **Bulk RNA-seq Preprocessing**: Downloading the full dataset and metadata using the SRA Toolkit and Athena, and setting up directories\n",
    "- **Quality Control**: Use FastQC and MultiQC to assess the quality of reads in the dataset and combine results for all samples to generate a comprehensive overview of quality metrics across multiple samples.\n",
    "- **Adapter Trimming**: Learn how to use Trimmomatic to remove adapter sequences and low-quality bases from FASTQ reads.\n",
    "- **Read Mapping and Quantification**: Understand the purpose of indexing and learn how to use STAR to create an index of the reference genome for efficient read mapping. Map reads to reference genome and quantify gene expression levels using RSEM.\n",
    "\n",
    "In summary, this Jupyter Notebook provided a hands-on demonstration of an extended Bulk RNA-Seq analysis workflow, guiding users through essential steps such as obtaining sequencing data from the Sequence Read Archive using the SRA Toolkit and Athena, read trimming with Trimmomatic, quality control with FastQC and MultiQC, reference genome indexing with STAR, read mapping and quantification using RSEM, and storing data on an AWS S3 bucket. By processing the full set of reads from a mouse dataset, we were able to observe the expression levels of the top 10 most highly expressed genes in each sample. This workflow serves as a foundation for more advanced analyses, and further resources are available for utilizing R/DESeq2 for differential gene expression analysis using the read counts generated in this tutorial, and using NetAct for transcription factor network analysis. Ultimately, this tutorial equips users with the basic skills to analyze RNA-seq data and to understand the core components of a typical RNA-seq pipeline.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a name=\"workflow\">Additional Workflows</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that you have read counts per gene, feel free to explore the R workflow which creates plots and analyses using these readcount files, or try other alternate workflows for creating read count files, such as using snakemake."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "[Workflow One:](Tutorial_1_subsampling_mouse-miniforge.ipynb) A short introduction to downloading and mapping sequences to a mouse genome using STAR and RSEM.\n",
    "\n",
    "\n",
    "[Workflow Two (DEG Analysis):](Tutorial_2_DEG_Analysis_mouse.ipynb) Using Deseq2 and R to conduct clustering and differential gene expression analysis.\n",
    "\n",
    "[Workflow Three (Network Analysis):](Tutorial_3_NetAct.ipynb) Using NetAct and R to conduct transcription factor network analysis.\n"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "conda-base-py",
   "name": "workbench-notebooks.m137",
   "type": "gcloud",
   "uri": "us-docker.pkg.dev/deeplearning-platform-release/gcr.io/workbench-notebooks:m137"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
